src/weaponstats.cc

```diff
@@ -0,0 +1,171 @@
+#include "weaponstats.hh"
+#include "functions.hh"
+#include <cstdlib>
+#include <algorithm>
+#include "injector/injector.hpp"
+#include "logger.hh"
+#include "base.hh"
+#include <iostream>
+#include "config.hh"
+
+int                    statsStart                       = 0xC8AAB8;
+int                    itemSize                         = 0x70;
+WeaponStatsRandomizer *WeaponStatsRandomizer::mInstance = nullptr;
+int                    hexFlags[]
+    = {0x3033, 0x3833, 0x7001, 0x7011,  0x7031,  0x2001, 0x2011, 0x2031, 0x3013,
+       0x3033, 0xA008, 0xA014, 0x48214, 0x30238, 0x238,  0x0038, 0x0431};
+
+/*******************************************************/
+int __fastcall RandomizeStats (int *address, int weaponid)
+{
+    CPed *player = FindPlayerPed ();
+
+    if (address < (int *) player || address > (int *) (player + 0x764)
+        || weaponid == 41 || weaponid == 42 || weaponid == 43)
+        return 0;
+
+    int weaponids[] = {weaponid, 0, 0, 0};
+
+    if (weaponid + 25 >= 47 && weaponid + 25 <= 57)
+        {
+            weaponids[1] = weaponid + 25;
+            weaponids[2] = weaponid + 36;
+            weaponids[3] = weaponid + 47;
+        }
+
+    short quality = random (100);
+
+    short ammoClip;
+    short damage;
+    float accuracy;
+    float moveSpeed;
+    float range;
+    int   properties;
+
+    if (weaponid == 18 || weaponid == 16 || weaponid == 39 || weaponid == 17)
+        {
+            properties = 0x100;
+        }
+    else if (weaponid == 35 || weaponid == 36)
+        {
+            properties = 0x48214;
+        }
+    else if (weaponid == 38)
+        {
+            properties = 0x238;
+        }
+    else
+        {
+            properties = hexFlags[random (0, 16)];
+        }
+
+    if (quality < 10)
```

Comment:
So I don't quite understand the quality part, isn't it much more likely for the weapon to be low quality than for it to be high quality? Since you only change the lower bounds for the randomness.

Look at the ammo-clip for example,
- Poor Quality: 1 to 15
- Okay Quality: 1 to 30
- Medium Quality: 1 to 50
- High Quality: 1 to 100

So as you see, there's a much higher chance for a lower number, since you have to get **both** a high quality number, **and** a high random damage number later. Perhaps, something like this would make more sense?

- 1 to 15
- 15 to 30
- 30 to 50
- 50 to 100

Again, I don't know if this was intentional, so you can ignore this if this is what you wanted.

Reply:
The main reason I created quality is that I wanted to have a chance to get really good weapons while the game was still beatable. When I was testing I found that Just Business is impossible to finish when the SMG damage values got really high. So I created this situation when you have small possibility to roll good weapon rather than keeping situation when all weapons have pretty mediocre damage.